tests/test_automatic_fields.py

```diff
@@ -15,38 +21,44 @@ def test_run(self):
         expected_streams = self.expected_streams()
 
-        # run check mode
+        # Run check mode
         found_catalogs = self.run_and_verify_check_mode(conn_id)
 
-        # de-select all the fields
-        self.select_found_catalogs(conn_id, found_catalogs, deselect_all_fields=True, only_streams=expected_streams)
+        # De-select all the fields
+        self.select_found_catalogs(conn_id, found_catalogs,
+                                   deselect_all_fields=True, only_streams=expected_streams)
 
-        # run sync
+        # Run sync
         record_count_by_stream = self.run_and_verify_sync(conn_id)
         synced_records = runner.get_records_from_target_output()
 
         for stream in expected_streams:
             with self.subTest(stream=stream):
 
-                # expected values
+                # Expected values
                 expected_primary_keys = self.expected_primary_keys()[stream]
                 expected_keys = expected_primary_keys | self.expected_replication_keys()[stream]
 
-                # collect actual values
+                # Collect actual values
                 messages = synced_records.get(stream)
-                record_messages_keys = [set(row['data'].keys()) for row in messages['messages']]
+                record_messages_keys = [set(row['data'].keys())
+                                        for row in messages['messages']]
 
-                # check if the stream has collected some records
+                # Check if the stream has collected some records
                 self.assertGreater(record_count_by_stream.get(stream, 0), 0)
 
                 # Verify that only the automatic fields are sent to the target
                 for actual_keys in record_messages_keys:
                     self.assertSetEqual(expected_keys, actual_keys)
 
+                # Get records
+                records = [message.get("data") for message in messages.get('messages', [])
+                           if message.get('action') == 'upsert']
+                # Remove duplicate records
+                records_pks_list = [tuple(message.get(pk) for pk in expected_primary_keys)
+                                   for message in [json.loads(t) for t in {json.dumps(d) for d in records}]]
+                records_pks_set = set(records_pks_list)
+
                 # Verify we did not duplicate any records across pages
-                records_pks_set = {tuple([message.get('data').get(primary_key) for primary_key in expected_primary_keys])
-                                          for message in messages.get('messages')}
-                records_pks_list = [tuple([message.get('data').get(primary_key) for primary_key in expected_primary_keys])
-                                           for message in messages.get('messages')]
                 self.assertCountEqual(records_pks_set, records_pks_list,
                                       msg="We have duplicate records for {}".format(stream))
```

Comment:
On L#58 we are removing duplicate records and then collecting the primary keys so basically there were duplicate records before but now there are none. That means length of primary keys list is always going to be equal to length of primary keys set. 

After the checking a API response for `tasks` stream, we confirmed that API returns the duplicate records in the same paginated response itself. so we can have duplicate records. Considering that this code looks redundant and should be removed. 

On the other, same code can check if we can have different records with same primary key, implying that defined primary key is not a valid primary key. I would go with this approach and change the error msg on L#64.

Reply:
Updated the assertion msg as well as the comment to describe the scenario.